File: coauthor/agent/model_factory.py
"""Factory for creating model handlers."""

from .model_config import ModelConfig, ModelProvider
from .model_handler import ModelHandler
from .model_handler_openai import OpenAIHandler
from .model_handler_anthropic import AnthropicHandler
from .model_handler_google import GoogleviaOpenAIHandler
from .model_handler_openrouter import OpenRouterHandler, AnthropicviaOpenrouterHandler


class ModelFactory:
    """Factory for creating model handlers with appropriate handlers."""

    @staticmethod
    def create_handler(config: ModelConfig) -> ModelHandler:
        """Create model handler based on provider and OpenRouter configuration."""
        # Handle OpenRouter configuration
        if config.useOpenRouter:
            # Set OpenRouter model name if not provided
            if not config.openrouterFullName:
                config.openrouterFullName = f"{config.provider.value}/{config.fullName}"

            # Route to appropriate OpenRouter handler
            if config.provider == ModelProvider.ANTHROPIC:
                return AnthropicviaOpenrouterHandler(config)
            return OpenRouterHandler(config)

        # Map providers to their handler classes
        handler_map = {
            ModelProvider.ANTHROPIC: AnthropicHandler,
            ModelProvider.OPENAI: OpenAIHandler,
            ModelProvider.GOOGLE: GoogleviaOpenAIHandler,
            ModelProvider.OTHERS: OpenRouterHandler,
        }

        handler_class = handler_map.get(config.provider)
        if not handler_class:
            raise ValueError(f"Unsupported model provider: {config.provider}")

        return handler_class(config)

File: coauthor-for-vs-code/src/agent/ModelFactory.ts
// Local imports - agent components
import { ModelConfig, ModelProvider } from './ModelConfig';
import { ModelHandler } from './ModelHandler';
import { ModelHandlerAnthropic } from './modelHandlerAnthropic';
import { ModelHandlerGoogle } from './modelHandlerGoogle';
import {
  ModelHandlerOpenRouter,
  ModelHandlerAnthropicViaOpenRouter,
} from './modelHandlerOpenRouter';
import { ModelHandlerOpenAI } from './modelHandlerOpenAI';

/** Factory class for instantiating appropriate model handlers based on configuration. */
export class ModelFactory {
  /**
   * Creates a model handler instance based on provider and routing configuration.
   * @param config Model configuration including provider and OpenRouter settings
   * @returns Appropriate ModelHandler instance for the configuration
   * @throws Error if provider is unsupported
   */
  static createHandler(config: ModelConfig): ModelHandler {
    // Handle OpenRouter configuration
    if (config.useOpenRouter) {
      // Set OpenRouter model name if not provided
      if (!config.openrouterFullName) {
        config.openrouterFullName = `${config.provider}/${config.fullName}`;
      }

      // Route to appropriate OpenRouter handler
      if (config.provider === ModelProvider.ANTHROPIC) {
        return new ModelHandlerAnthropicViaOpenRouter(config);
      }
      return new ModelHandlerOpenRouter(config);
    }

    // Map providers to their handler classes
    const handlerMap = new Map<
      ModelProvider,
      new (config: ModelConfig) => ModelHandler
    >([
      [ModelProvider.ANTHROPIC, ModelHandlerAnthropic],
      [ModelProvider.OPENAI, ModelHandlerOpenAI],
      [ModelProvider.GOOGLE, ModelHandlerGoogle],
      [ModelProvider.OTHERS, ModelHandlerOpenRouter],
    ]);

    const HandlerClass = handlerMap.get(config.provider);
    if (!HandlerClass) {
      throw new Error(`Unsupported model provider: ${config.provider}`);
    }

    return new HandlerClass(config);
  }
}
